article/repository/mysql_article.go

```diff
@@ -213,3 +215,26 @@ func EncodeCursor(t time.Time) string {
 
 	return base64.StdEncoding.EncodeToString([]byte(timeString))
 }
+
+func (m *mysqlArticleRepository) WithTransaction(f func(repo article.Repository) error) error {
```

Comment:
Hi @philchia , thanks for the PR. 

I have a question, have you tried with context?

I have another approachment, 
I'm using context, then insert the `tx` to Context. 

So generally, my function will look like this, 

```go
func (m * mysqlArticleRepository) WithTransaction(ctx context.Context, fn func(c context.Context) error) (err error) {
	tx, err := m.DB.BeginTx(ctx, nil)
	if err != nil {
		return
	}

	_, err = tx.ExecContext(ctx, "SET TRANSACTION ISOLATION LEVEL READ COMMITTED")
	if err != nil {
		return
	}

	c := context.WithValue(ctx, "TransactionContextKey", tx)
	err = fn(c)
	if err != nil {
		if errTX := tx.Rollback(); errTX != nil {
			logrus.Error("failed to rollback transaction", errTX)
		}
		return
	}

	if errTX := tx.Commit(); errTX != nil {
		logrus.Error("failed to commmit transaction", errTX)
	}
	return
}
```

Usage in usecase/service:
```go
err = s.repo.WithTransaction(ctx, func(c context.Context) (err error) {
		err = s.repo.Create(c, article)
		if err != nil {
			return
		}
		err = s.authorRepo.Create(c, author)
		return
	})
```

But, the cons, maybe a bit redundant jobs, we need to check the `tx` in ctx. 
Like this. 

```go
func getTransactionFromCtx(ctx context.Context) (*sql.Tx, bool) {
	tx, ok := ctx.Value("TransactionContextKey").(*sql.Tx)
	return tx, ok
}

func (m *mysqlArticleRepository) Insert(ctx context.Context, article Article)error{
	var tx *sql.Tx
	tx, isTxFromCtx := getTransactionFromCtx(ctx)
	if !isTxFromCtx {
		tx, err = r.DB.BeginTx(ctx, nil)
		if err != nil {
			return
		}
	}
        tx.ExecContext(ctx, "INSERT article SET title=?", article.Titlle)
//....
}
```

Reply:
Hi, thanks for this architecture.
But I would like to object to the use of context to convey something not explicitly.
I think it’s worthwhile either to wrap the transaction (which is suggested in this PR), or to transfer the transaction explicitly.
Also, as an option, it may just be worth creating a method in which there will be a transaction inside (for example CreateArticleWithAuthor).

Reply:
hi @ghostiam,
Have you tried with the submitted PR, is there any issue so far?
Maybe related to memory and performance?

Or anything?

Reply:
Transferring transaction explicitly is something I don't want to do ideally. 

Since, the use case layer (who will use the repository), shouldn't know anything about the transaction. It's only know about calling repository. 

What if, the repository using NoSQL, which is doesn't support transactions?
Let's say like MongoDB in the early stage, that still doesn't support transaction?

Reply:
There are plenty of possible business reasons why you would need to batch multiple repositories into one transaction though. By not passing the `tx` in explicitly, you will prevent the developer from being able to do this.

NoSQL and SQL databases are pretty different, if you want to have both persistence layers be supported by the repository pattern, you'll only be able to use the features that are available to both types of databases, meaning you won't be able to support transactions.

Reply:
Hi John @nii236
So in your opinion should we pass the `tx` to usecase layer here?

Reply:
That's correct, transactions should live on a layer above a repository, because they are often used across multiple repositories.

Reply:
@nii236 I also agree here. But this "violates" the "clean" architecture since the use case layer knows about the implementation details.

Question is what the "clean" architecture recommends for handling transactions in a RDBMS?

Is there something that you do not have to reinvent the wheel?

Reply:
Hello all, i just read the discussion. 

I have little case, let's say i use 1 gRPC handling with calling 2 or more use case inside. 

where should transaction place? imagine 1 gRPC handler do Create order and Create user at same time (not same really). It call first usecase and the result will be use for call second usecase. 

what happen when the second usecase fail? but the data store from firts usecase is successful. 
maybe transaction is needed but still confused where to put it.

Reply:
> @nii236 I also agree here. But this "violates" the "clean" architecture since the use case layer knows about the implementation details.

It can be done without braking the "clean" rule, if you a abstract a interface for repositories that has `Begin()`, `Commit()`, `Rollback()`.
Doing this, the use case layer won't know any details about how repositories implement, and still has the power to control transaction.

Reply:
My 2cents. I fully agree with @bxcodec. Usecases should not know about transactions at all. Tx are a repo thing.

Reply:
@henrylusjen do you have an example, a gist?

Reply:
> My 2cents. I fully agree with @bxcodec. Usecases should not know about transactions at all. Tx are a repo thing.

I want to point out that we are talking about two different kinks of the transaction. One is the "DB" transaction, another one is business logic transaction (multiple work should do together).

Yes, usecase should not know about "DB" transaction at all. But like @nii236 said, it should have the power to raise "Business" transaction, and it's much complicated because it usually involve multiple repos and sometimes MQ or external service. 

There's a solution called "UnitOfWork", but golang doesn't have any ORM support it yet. You might have to build it yourself.

Reply:
OK. Understood. Thanks.
What do you think about https://github.com/frederikhors/go-clean-arch/commit/2ec0f2d7d8d4a5f5ed076e4a1ce59eb862333ca7?